system/Entity.php

```diff
@@ -527,69 +508,67 @@ protected function mutateDate($value)
 	 * Provides the ability to cast an item as a specific data type.
 	 * Add ? at the beginning of $type  (i.e. ?string) to get NULL instead of casting $value if $value === null
 	 *
-	 * @param mixed  $value
-	 * @param string $type
+	 * @param mixed  $value     Attribute value
+	 * @param string $attribute Attribute name
+	 * @param string $method    Allowed to "get" and "set"
 	 *
 	 * @return mixed
 	 * @throws Exception
 	 */
-	protected function castAs($value, string $type)
+	protected function castAs($value, string $attribute, string $method)
 	{
+		if (empty($this->casts[$attribute]))
+		{
+			return $value;
+		}
+
+		$type = $this->casts[$attribute];
+
 		if (strpos($type, '?') === 0)
 		{
 			if ($value === null)
 			{
 				return null;
 			}
+
 			$type = substr($type, 1);
 		}
 
-		switch($type)
+		//In order not to create a separate handler for the json-array type, we transform the required one.
+		$type = $type === 'json-array' ? 'json[array]' : $type;
+
+		if (! in_array($method, ['get', 'set'], true))
 		{
-			case 'int':
-			case 'integer': // alias for 'integer'
-				$value = (int) $value;
-				break;
-			case 'float':
-				$value = (float) $value;
-				break;
-			case 'double':
-				$value = (double) $value;
-				break;
-			case 'string':
-				$value = (string) $value;
-				break;
-			case 'bool':
-			case 'boolean': // alias for 'boolean'
-				$value = (bool) $value;
-				break;
-			case 'csv':
-				$value = explode(',', $value);
-				break;
-			case 'object':
-				$value = (object) $value;
-				break;
-			case 'array':
-				if (is_string($value) && (strpos($value, 'a:') === 0 || strpos($value, 's:') === 0))
-				{
-					$value = unserialize($value);
-				}
+			throw CastException::invalidCastMethod();
+		}
+
+		$params = [];
+
+		//Attempt to retrieve additional parameters if specified
+		// type[param, param2,param3]
+		if (preg_match('/^(.+)\[(.+)\]$/', $type, $matches))
+		{
+			$type   = $matches[1];
+			$params = array_map(function ($param) {
+				return trim($param);
+			}, explode(',', $matches[2]));
+		}
+
```

Comment:
We should have `nullable` added in parameters `if (strpos($type, '?') === 0)`

There might be a situation where `null` might have to be return instead of $value. For example casting empty string to `?int`. The cast method should return null in such case but having `int` as cast should trow an exception or maybe in some custom cast class return 0.

Reply:
The nullable type is handled above.
I see no reason to apply nullable to parameters. They are not computed and are passed to the handler as a string

Reply:
There are situations where it might be useful. There is no way to submit null via form. So in most cases I have custom setters which will set the empty string got via post to null. 

Having nullable as parameter in custom setter object will make it possible to completely get rid of such setters as it will be possible to make custom cast which will handle empty string to null conversion.

Reply:
How do you imagine that? I don't understand.
'field' => 'my_type[?param]'  Like this?

Reply:
set $isNullable inside if (strpos($type, '?') === 0)

And then here after pregmatch append it to the params

if($isNullable)
{
  $params[] = ‘nullable’;
}

Reply:
That's a good question. It can be done that way but it makes things very confusing as there will be 3 cases which allow null values.

`?bool`
`?bool[nullable]`
`bool[nullable]`

Reply:
OK. Commit with your suggestion added